text/0000-cargo-embed-dependency-versions.md

```diff
@@ -0,0 +1,197 @@
+- Feature Name: `cargo_embed_dependency_versions`
+- Start Date: 2019-11-03
+- RFC PR: [rust-lang/rfcs#2801](https://github.com/rust-lang/rfcs/pull/2801)
+- Rust Issue: None
+
+# Summary
+[summary]: #summary
+
+Embed the crate dependency tree in a machine-readable format into compiled binaries so it could be programmatically recovered later.
+
+# Motivation
+[motivation]: #motivation
+
+Rust is very promising for security-critical applications due to its safety guarantees, but there currently are gaps in the ecosystem that prevent it. One of them is the lack of any infrastructure for security updates.
+
+Linux distributions alert you if you're running a vulnerable software version and you can opt in to automatic security updates. Cargo not only has no automatic update infrastructure, it doesn't even know which libraries or library versions went into compiling a certain binary, so there's no way to check if your system is vulnerable or not.
+
+The primary use case for this information is cross-referencing versions of the dependencies against [RustSec advisory database](https://github.com/RustSec/advisory-db) and/or third-party databases such as [Common Vulnerabilities and Exposures](https://en.wikipedia.org/wiki/Common_Vulnerabilities_and_Exposures). This also enables use cases such as ensuring a fix in a depencency has been propagated across the entirety of your fleet or preventing binaries with unvetted dependencies from accidentally reaching a production environment - all with zero bookkeeping.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Every time an executable is compiled with Cargo, the dependency tree of the executable is recorded in the binary. This includes the names, versions, dependency kind (build or normal), and origin kind (crates.io, git, local filesystem, custom registry). Development dependencies are not recorded, since they cannot affect the final binary. All filesystem paths and URLs are redacted to preserve privacy. The data is encoded in JSON and compressed with zlib to reduce its size.
+
+This data can be recovered using existing tools like `readelf` or Rust-specific tooling. It can be then used to create a Software Bill of Materials in a common format, or audit the dependency list for known vulnerabilities.
+
+WASM, asm.js and embedded platforms are exempt from this mechanism by default since they have very strict code size requirements. For those platforms we encourage you to use tooling that record the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.
+
+A per-profile configuration option in `Cargo.toml` can be used to opt out of this behavior if it is not desired (e.g. when building [extremely minimal binaries](https://github.com/johnthagen/min-sized-rust)). The exact name of this option is subject to bikeshedding.
```

Comment:
```suggestion
Every time an executable is compiled with Cargo, Cargo can record the dependency tree of the executable in a special section in the binary. This includes the names, versions, dependency kind (build or normal), and origin kind (crates.io, git, local filesystem, custom registry). Development dependencies are not recorded, since they cannot affect the final binary. All filesystem paths and URLs are redacted to preserve privacy. The data is encoded in JSON and compressed with zlib to reduce its size.

This data can be recovered using existing tools like `readelf` or Rust-specific tooling. It can then be used to create a Software Bill of Materials (SBOM) in a common format, or audit the dependency list for known vulnerabilities.

A per-profile configuration option in `Cargo.toml` can be used to opt into this behavior when desired (e.g. when not concerned about binary size or information disclosure). The exact name of this option is subject to bikeshedding.

For users who don't enable this option, we encourage using tooling that separately records the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.
```

(And perhaps an addition to the future-work section saying that Cargo itself could help with the separate-database case, perhaps by emitting the dep information in a separate file rather than embedded, so that the user can record it.)

Toggling this from opt-out to opt-in.